crates/pallet-domains/src/block_tree.rs

```diff
@@ -93,24 +109,34 @@ pub(crate) fn execution_receipt_type<T: Config>(
         Ordering::Greater => ReceiptType::Rejected(RejectedReceiptType::InFuture),
         Ordering::Equal => ReceiptType::Accepted(AcceptedReceiptType::NewHead),
         Ordering::Less => {
+            // Reject receipt that already pruned/confirmed
             let oldest_receipt_number =
                 head_receipt_number.saturating_sub(T::BlockTreePruningDepth::get());
-            let already_exist = BlockTree::<T>::get(domain_id, receipt_number)
-                .contains(&execution_receipt.hash::<DomainHashingFor<T>>());
-
             if receipt_number < oldest_receipt_number {
-                // Receipt already pruned
-                ReceiptType::Rejected(RejectedReceiptType::Pruned)
-            } else if !already_exist {
-                // Create new branch
-                ReceiptType::Accepted(AcceptedReceiptType::NewBranch)
-            } else if receipt_number == head_receipt_number {
-                // Add comfirm to the current head receipt
-                ReceiptType::Accepted(AcceptedReceiptType::CurrentHead)
-            } else {
-                // Add comfirm to a non-head receipt
-                ReceiptType::Stale
+                return ReceiptType::Rejected(RejectedReceiptType::Pruned);
+            }
+
+            // Reject receipt that try to create new branch in the block tree
+            let already_exist = is_receipt_exist::<T>(
+                domain_id,
+                receipt_number,
+                execution_receipt.hash::<DomainHashingFor<T>>(),
+            );
+            if !already_exist {
+                return ReceiptType::Rejected(RejectedReceiptType::NewBranch);
             }
+
+            // Add confirm to the head receipt that added in the current block or it is
+            // the genesis receipt
+            let head_receipt_extended = HeadReceiptExtended::<T>::get(domain_id);
+            if receipt_number == head_receipt_number
+                && (head_receipt_extended || receipt_number.is_zero())
```

Comment:
for `receipt_number.is_zero()` we should ensure the `head_receipt_number` is One

Reply:
Not really, `receipt_number.is_zero()` is a special handling for the first bundle in the first domain block, and we should ensure `head_receipt_number` is zero, which is done by the `receipt_number == head_receipt_number` condition.

Reply:
ahh I see. okay